### ISSUE TITLE
Large asset upload timed out.

### ISSUE THREAD

**Issue by tokejepsen** [PyGithub/PyGithub#693]
Trying to upload an asset of ~600 MB to a GitHub release, but getting the below error.

```python
Traceback (most recent call last):
  File "run_deploy.py", line 47, in <module>
    label=release_title
  File "C:\Python27-x64\lib\site-packages\github\GitRelease.py", line 161, in upload_asset
    input=path
  File "C:\Python27-x64\lib\site-packages\github\Requester.py", line 177, in requestBlobAndCheck
    return self.__check(*self.requestBlob(verb, url, parameters, headers, input))
  File "C:\Python27-x64\lib\site-packages\github\Requester.py", line 245, in requestBlob
    return self.__requestEncode(None, verb, url, parameters, headers, input, encode)
  File "C:\Python27-x64\lib\site-packages\github\Requester.py", line 268, in __requestEncode
    status, responseHeaders, output = self.__requestRaw(cnx, verb, url, requestHeaders, encoded_input)
  File "C:\Python27-x64\lib\site-packages\github\Requester.py", line 295, in __requestRaw
    response = cnx.getresponse()
  File "C:\Python27-x64\lib\httplib.py", line 1121, in getresponse
    response.begin()
  File "C:\Python27-x64\lib\httplib.py", line 438, in begin
    version, status, reason = self._read_status()
  File "C:\Python27-x64\lib\httplib.py", line 394, in _read_status
    line = self.fp.readline(_MAXLINE + 1)
  File "C:\Python27-x64\lib\socket.py", line 480, in readline
    data = self._sock.recv(self._rbufsize)
  File "C:\Python27-x64\lib\ssl.py", line 766, in recv
    return self.read(buflen)
  File "C:\Python27-x64\lib\ssl.py", line 653, in read
    v = self._sslobj.read(len)
ssl.SSLError: ('The read operation timed out',)
```

All I'm doing is ```release.upload_asset("/path/to/deployment.zip")```.

Am I using the library wrong somehow?

**Comment by ncapalbo** [comment]
Within the last week, I've been encountering the same SSL error (i.e. 'The read operation timed out') with frequency I've never experienced before.  The API call that seems to be causing it is get_public_events() of the NamedUser object.  I've tried a bunch of things to reproduce the error but I can't pinpoint the root cause.

**Comment by sfdye** [comment]
Is this still an issue? 

@ncapalbo can you provide some details of your case?

**Comment by ncapalbo** [comment]
@sfdye apologies for the delay in following up to this!

I'm still experiencing intermittent failures in using the library, but it's very unclear what exactly is the root cause (i.e. whether it's PyGitHub or a strange behavior since the script is running in AWS Lambda).  After a bit of debugging here is the snippet of code where the issue is occurring:

```python
    while error_count < 4:
      try:
        results = self.find_sensitive_commits(users)
        break
      except Exception as e:
        error_count += 1
        self._logger.error(
            'Error #%d; unable to complete audit: %s' % (
                error_count, e.message))
```

find_sensitive_commits() looks like this:
```python
  def find_sensitive_commits(self, users):
    """Get all the public events for our org users and audit them.
    Args:
        users (list): a list of github.NamedUser objects.
    Returns:
        A dict containing all commits that contain sensitive keywords.
    """
    results = []
    for user in users:
      # TODO(ncapalbo): it is understood that this print statement is going
      # create quite a volume of logging.  The root cause of <redacted> is
      # still unknown, so this will possibly help debugging efforts.
      self._logger.info('Analyzing public commits of user "%s"' % user.login)
      events = [i for i in user.get_public_events() if i.type == 'PushEvent']
      for event in events:
        if self.is_expired_commit(event.created_at):
          continue
```

Print statement debugging has indicated the failure is occurring on the call to user.get_public_events().  What's slightly odd about this whole thing is when I catch the exception, there is no e.message:

```
[ERROR] 2018-03-19T04:34:34.811Z  498a7277-2b2e-11e8-823d-bb16359c081e  Error #2; unable to complete audit:
```

I thought maybe it was due to a particular commit from a particular GitHub user that consistently causes the script to crash, but the user prior to the error condition is not always the same.  I'm not sure if any of this helps understand the issue, but I'm happy to elaborate more if need be.  Thanks in advance for your help with this!

**Comment by sfdye** [comment]
Very interesting use case indeed! I am sorry but it seems the topic has diverged as the original issue was with the asset upload API and now is `get_public_events()`? What is the exact issue please?

**Comment by ncapalbo** [comment]
Apologies if my response wasn't clear.  The exact issue is (or at least was) that both calls -- get_public_events() as well as upload_asset() -- were resulting in the `The read operation timed out` SSL error from the original comment.  My error seems to be diverging from that now, but again, it's hard to tell since whatever exception is being caught doesn't apparently have a message object field.  Is that more clear?

**Comment by sfdye** [comment]
Thanks, I am clearer now. Give me some time to investigate the issue and get back to you. This might be a bit tricky to debug though.

**Comment by mfonville** [comment]
This is a notorious problem on the side of GitHub (the AWS infrastructure regularly fails) and the same happens with e.g. GotHub/GitHub-Release (https://github.com/itchio/gothub)

Best would be if at least pygithub can handle this situation gracefully, and that we can give examples to users how to deal with the situation (recognizing a failed upload, removing an incomplete upload asset, retrying)

**Comment by sfdye** [comment]
Oh I just saw your PR #768

**Comment by mfonville** [comment]
With PyGithub now using requests in the backend, the error has changed a little bit, and should be able to be caught by the developer if checking for HTTP status 502 in the GithubException.

```
  File "/home/maarten/PyGithub/build/lib/github/Requester.py", line 261, in requestBlobAndCheck
    return self.__check(*self.requestBlob(verb, url, parameters, headers, input, self.__customConnection(url)))
  File "/home/maarten/PyGithub/build/lib/github/Requester.py", line 266, in __check
    raise self.__createException(status, responseHeaders, output)
github.GithubException.GithubException: 502 {'message': 'Error uploading asset', 'request_id': 'xxxxxxxxxxxxxxxxxxxxxxx'}
```

If that happens, it is necessary to remove the partially uploaded asset before trying again; some example code:
```python
for asset in release.get_assets():
    if asset.name == filename and asset.state == "new":
        asset.delete_asset()
```

**Comment by mfonville** [comment]
An update, during my script I also got this one, a real http timeout:
```python
  File "/home/maarten/PyGithub/build/lib/github/GitRelease.py", line 210, in upload_asset
    input=path
  File "/home/maarten/PyGithub/build/lib/github/Requester.py", line 261, in requestBlobAndCheck
    return self.__check(*self.requestBlob(verb, url, parameters, headers, input, self.__customConnection(url)))
  File "/home/maarten/PyGithub/build/lib/github/Requester.py", line 266, in __check
    raise self.__createException(status, responseHeaders, output)
github.GithubException.GithubException: 504 {'data': '<!DOCTYPE html>\n<!--\n\nHello future GitHubber! I bet you\'re here to remove those nasty inline styles,\nDRY up these templates and make \'em nice and re-usable, right?\n\nPlease, don\'t. https://github.com/styleguide/templates/2.0\n\n-->\n<html> .....more HTML stuff here....
```
So a `504` is also a http status to check for.

**Comment by jhoblitt** [comment]
@mfonville I don't believe you are seeing the failure described at the root of this ticket as there would be no HTTP status code or json data with an error message.  Ultimately, the failure mode I see is a socket read error with both `1.39` and `1.40a3`, although the stack trace is different due to the changes with the plumbing.

`1.40a3`
---
```
Traceback (most recent call last):
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 387, in _make_request
    six.raise_from(e, None)
  File "<string>", line 2, in raise_from
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 383, in _make_request
    httplib_response = conn.getresponse()
  File "/usr/lib64/python3.6/http/client.py", line 1331, in getresponse
    response.begin()
  File "/usr/lib64/python3.6/http/client.py", line 297, in begin
    version, status, reason = self._read_status()
  File "/usr/lib64/python3.6/http/client.py", line 258, in _read_status
    line = str(self.fp.readline(_MAXLINE + 1), "iso-8859-1")
  File "/usr/lib64/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
  File "/usr/lib64/python3.6/ssl.py", line 965, in recv_into
    return self.read(nbytes, buffer)
  File "/usr/lib64/python3.6/ssl.py", line 827, in read
    return self._sslobj.read(len, buffer)
  File "/usr/lib64/python3.6/ssl.py", line 587, in read
    v = self._sslobj.read(len, buffer)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/requests/adapters.py", line 440, in send
    timeout=timeout
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 639, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/urllib3/util/retry.py", line 357, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/urllib3/packages/six.py", line 686, in reraise
    raise value
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 601, in urlopen
    chunked=chunked)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 389, in _make_request
    self._raise_timeout(err=e, url=url, timeout_value=read_timeout)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/urllib3/connectionpool.py", line 309, in _raise_timeout
    raise ReadTimeoutError(self, url, "Read timed out. (read timeout=%s)" % timeout_value)
urllib3.exceptions.ReadTimeoutError: HTTPSConnectionPool(host='api.github.com', port=443): Read timed out. (read timeout=10)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jhoblitt/github/sqre-codekit/venv/bin/github-fork-org", line 11, in <module>
    load_entry_point('sqre-codekit', 'console_scripts', 'github-fork-org')()
  File "/home/jhoblitt/github/sqre-codekit/codekit/cli/github_fork_org.py", line 377, in main
    dry_run=args.dry_run
  File "/home/jhoblitt/github/sqre-codekit/codekit/cli/github_fork_org.py", line 160, in create_teams
    dst_t = org.create_team(name, repo_names=repos)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/PyGithub-1.40a3-py3.6.egg/github/Organization.py", line 470, in create_team
    input=post_parameters
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/PyGithub-1.40a3-py3.6.egg/github/Requester.py", line 258, in requestJsonAndCheck
    return self.__check(*self.requestJson(verb, url, parameters, headers, input, cnx))
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/PyGithub-1.40a3-py3.6.egg/github/Requester.py", line 304, in requestJson
    return self.__requestEncode(cnx, verb, url, parameters, headers, input, encode)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/PyGithub-1.40a3-py3.6.egg/github/Requester.py", line 357, in __requestEncode
    status, responseHeaders, output = self.__requestRaw(cnx, verb, url, requestHeaders, encoded_input)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/PyGithub-1.40a3-py3.6.egg/github/Requester.py", line 384, in __requestRaw
    response = cnx.getresponse()
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/PyGithub-1.40a3-py3.6.egg/github/Requester.py", line 105, in getresponse
    r = verb(url, headers=self.headers, data=self.input, timeout=self.timeout, verify=self.verify)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/requests/sessions.py", line 555, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/home/jhoblitt/github/sqre-codekit/venv/lib/python3.6/site-packages/requests/adapters.py", line 521, in send
    raise ReadTimeout(e, request=request)
requests.exceptions.ReadTimeout: HTTPSConnectionPool(host='api.github.com', port=443): Read timed out. (read timeout=10)

```

**Comment by jhoblitt** [comment]
@sfdye It looks like the 10s default socket timeout, which I suspect is coming from `socket._GLOBAL_DEFAULT_TIMEOUT` (opaque object -- seems to hold no data) timeout is too short.  If I hack requests to force a 45s timeout, I can get to a more useful `github.GithubException.GithubException: 502 {'message': 'Server Error'}`.  Perhaps some sort of knob to allow end users to set a max read timeout would be reasonable?

**Comment by jhoblitt** [comment]
<facepalm/> It looks like the default is coming from `github.MainClass.DEFAULT_TIMEOUT`...

**Comment by mfonville** [comment]
Here is a `npm` tool to make GitHub releases, and they also had to increase the timeout of their socket it seems: https://github.com/webpro/release-it/issues/50

I do understand btw that with the old plumbing there would be a socket-read error, because it is known that `httplib` is known not to be able to detect a closed socket (I read in other github-api projects with similar issues) when the Github-server answers (while the client is sending the data) that it actually closing the socket and asks the client not to send anymore data; but I had thought/hoped that `requests` would be able to handle this situation better.

**Comment by mfonville** [comment]
maybe specifying another DEFAULT_TIMEOUT for uploads.github.com might be also a good starting point

**Comment by jhoblitt** [comment]
Something seems fishy on the github head.  I enabled debugging and cut'n'pasted the team creationg JSON causing the HTTP 502 via `requests` into a file to test `POST`ing it with`curl`.  Initially, it seemed reliable if I used `application/x-www-form-urlencoded` instead of `application/json` but both are now failing.  Sometimes the team is created on the 502 and some times it is not.  I think I'm going to ping GH support.

**Comment by jhoblitt** [comment]
I sent a `bash/curl` minimal example to support along with some request IDs.  The explanation I received is that there is a hard 10s limit for API requests to complete.  I suspect this is from a middle/back end and there is a bit of latency added by the front end (and network delay product), which pushes the max time for a response to be slightly beyond the pygithub default of 10s. Support also said they will look into HTTP 502s being returned even though the request is ultimately successful.  I did not, but should have, asked if the same timeout applies to asset uploads -- I just sent a follow up inquiry.

I suspect the best resolution is in change the pygithub default to something like 15s.

**Comment by jhoblitt** [comment]
I've opened #793 to bump the default timeout to 15s.

**Comment by sfdye** [comment]
I am fine with increasing the default timeout to 15s. But we still need to handle the asset upload timeout right? Which is much higher than normal api.

**Comment by mfonville** [comment]
Today I hit the broken pipe during upload:

```
('Connection aborted.', BrokenPipeError(32, 'Broken pipe'))
Traceback (most recent call last):
  File "/home/maarten/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 601, in urlopen
    chunked=chunked)
  File "/home/maarten/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 357, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1065, in _send_output
    self.send(chunk)
  File "/usr/lib/python3.6/http/client.py", line 986, in send
    self.sock.sendall(data)
  File "/usr/lib/python3.6/ssl.py", line 972, in sendall
    v = self.send(byte_view[count:])
  File "/usr/lib/python3.6/ssl.py", line 941, in send
    return self._sslobj.write(data)
  File "/usr/lib/python3.6/ssl.py", line 642, in write
    return self._sslobj.write(data)
BrokenPipeError: [Errno 32] Broken pipe


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/maarten/.local/lib/python3.6/site-packages/requests/adapters.py", line 440, in send
    timeout=timeout
  File "/home/maarten/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 639, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/home/maarten/.local/lib/python3.6/site-packages/urllib3/util/retry.py", line 357, in increment
    raise six.reraise(type(error), error, _stacktrace)
  File "/home/maarten/.local/lib/python3.6/site-packages/urllib3/packages/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/maarten/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 601, in urlopen
    chunked=chunked)
  File "/home/maarten/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 357, in _make_request
    conn.request(method, url, **httplib_request_kw)
  File "/usr/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/usr/lib/python3.6/http/client.py", line 1065, in _send_output
    self.send(chunk)
  File "/usr/lib/python3.6/http/client.py", line 986, in send
    self.sock.sendall(data)
  File "/usr/lib/python3.6/ssl.py", line 972, in sendall
    v = self.send(byte_view[count:])
  File "/usr/lib/python3.6/ssl.py", line 941, in send
    return self._sslobj.write(data)
  File "/usr/lib/python3.6/ssl.py", line 642, in write
    return self._sslobj.write(data)
urllib3.exceptions.ProtocolError: ('Connection aborted.', BrokenPipeError(32, 'Broken pipe'))



During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/maarten/PyGithub/build/lib/github/GitRelease.py", line 210, in upload_asset
    input=path
  File "/home/maarten/PyGithub/build/lib/github/Requester.py", line 261, in requestBlobAndCheck
    return self.__check(*self.requestBlob(verb, url, parameters, headers, input, self.__customConnection(url)))
  File "/home/maarten/PyGithub/build/lib/github/Requester.py", line 342, in requestBlob
    return self.__requestEncode(cnx, verb, url, parameters, headers, input, encode)
  File "/home/maarten/PyGithub/build/lib/github/Requester.py", line 365, in __requestEncode
    status, responseHeaders, output = self.__requestRaw(cnx, verb, url, requestHeaders, encoded_input)
  File "/home/maarten/PyGithub/build/lib/github/Requester.py", line 389, in __requestRaw
    response = cnx.getresponse()
  File "/home/maarten/PyGithub/build/lib/github/Requester.py", line 104, in getresponse
    r = verb(url, headers=self.headers, data=self.input, timeout=self.timeout)
  File "/home/maarten/.local/lib/python3.6/site-packages/requests/sessions.py", line 555, in post
    return self.request('POST', url, data=data, json=json, **kwargs)
  File "/home/maarten/.local/lib/python3.6/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/home/maarten/.local/lib/python3.6/site-packages/requests/sessions.py", line 618, in send
    r = adapter.send(request, **kwargs)
  File "/home/maarten/.local/lib/python3.6/site-packages/requests/adapters.py", line 490, in send
    raise ConnectionError(err, request=request)
```

**Comment by mfonville** [comment]
I think the asset timeout might also benefit from the 15 seconds, because as far as I know the server (often?) sends a response when it is closing the connection. I hopefully assume that during the upload, while it still having success, it would use a method to keep it alive and not letting the 10 second time-out kick in; and that maybe that response was also just outside the 10 second window.

In short: let's upgrade all the timeouts by default to 15 seconds and let's see what happens

**Comment by sfdye** [comment]
Released `1.40a4` containing fix from #793 

Please have a try and let me know how it works out.

**Comment by sfdye** [comment]
Feel free to re-open if still an issue.
